pmlc/conversion/tile_to_linalg/utils.cc

```diff
@@ -139,4 +141,118 @@ llvm::SmallSet<int64_t, 4> getUsedDims(AffineExpr expr) {
   return visitor.getUsedDims();
 }
 
+OpMapsAndShapes::OpMapsAndShapes(tile::ContractionOp op, ValueRange inputs,
+                                 ValueRange outputs, ArrayRef<AffineMap> maps)
+    : maps(maps), numDims(maps[0].getNumDims()) {
+  MLIRContext *context = op.getContext();
+  SmallVector<int64_t> lowerBounds =
+      op.lowerBounds().getValue().getConstantResults();
+  SmallVector<int64_t> upperBounds =
+      op.upperBounds().getValue().getConstantResults();
+  for (unsigned i = 0; i < lowerBounds.size(); i++) {
+    shape.emplace_back(upperBounds[i] - lowerBounds[i] + 1);
+  }
+  operands.insert(operands.end(), inputs.begin(), inputs.end());
+  operands.insert(operands.end(), outputs.begin(), outputs.end());
+  assert(operands.size() == maps.size());
+}
+
+// This function determines if all the loop dims appear as a single dim in
+// shape dims. If not, we need a dummp map to indicate the loop ranges.
+bool OpMapsAndShapes::needDummyMap() {
+  llvm::SmallSet<unsigned, 4> dims;
+  for (AffineMap map : maps) {
+    assert(numDims == map.getNumDims() &&
+           "The input maps have different numbers of dimensions.");
+    for (AffineExpr expr : map.getResults()) {
+      if (auto dimExpr = expr.dyn_cast<AffineDimExpr>()) {
+        dims.insert(dimExpr.getPosition());
+      }
+    }
+  }
+  return dims.size() != numDims;
+}
+
```

Comment:
Isn't this the same as:

https://github.com/llvm/llvm-project/blob/main/mlir/lib/IR/AffineMap.cpp#L514

Reply:
We can determine it using inversePermutation().